{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d46384-4849-4b4e-8950-4f248a7f0b5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Loads the .vcf files (Snakemake output) and generates a annotated rds dataset\n",
    "\n",
    "Requires:\n",
    "- pipeline_data/vcf_wgs_130721/*.vcf\n",
    "- datasets/contig_mapping_240621.csv\n",
    "- datasets/gff3_from_genbank_220621/*.gff3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6f97e70c-9493-40b1-be75-03fd4c42b290",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"../../\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4222ba95-d5e2-4b40-a8d7-f03addff9bd4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "also installing the dependencies ‘permute’, ‘cluster’, ‘ape’, ‘memuse’, ‘pinfsc50’, ‘vegan’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"vcfR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "79e79658-3cf9-4b42-a5e9-5acfe23c45f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.3 (2022-03-10)\n",
      "\n",
      "Installing package(s) 'BiocVersion', 'rtracklayer'\n",
      "\n",
      "also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘matrixStats’, ‘futile.logger’, ‘snow’, ‘MatrixGenerics’, ‘Biobase’, ‘DelayedArray’, ‘GenomeInfoDbData’, ‘BiocParallel’, ‘Rhtslib’, ‘SummarizedExperiment’, ‘rjson’, ‘GenomicRanges’, ‘XML’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘GenomeInfoDb’, ‘Biostrings’, ‘zlibbioc’, ‘Rsamtools’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'blob', 'gert', 'later', 'parallelly', 'randomForest', 'RSQLite'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(\"rtracklayer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8db1ee4e-28df-4ae8-a7fe-057d065f5034",
   "metadata": {},
   "outputs": [],
   "source": [
    "#source(\"utils/helpers.R\")\n",
    "library(rtracklayer)\n",
    "library(data.table)\n",
    "library(tidyr)\n",
    "library(vcfR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "08bbf46a-d01c-4a2d-8874-2b4c285206a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(OligoMMR2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "674e2a60-e16b-4f36-b1d8-b9272011d945",
   "metadata": {},
   "source": [
    "Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "218b599b-08f5-4b35-8d83-fac134a636a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "qual_threshold <- 0 # removes SNPs that have a value below qual_threshold in the QUAL field in the .vcf file\n",
    "vcf_files <- Sys.glob(\"pipeline_data/vcf/vcf_wgs_130721/*.vcf\") # Path to vcf files\n",
    "gff_files <- Sys.glob(\"datasets/gff3_from_genbank_220621/*.gff3\") # Path to annotaiton of reference genomes\n",
    "contig_mapping_path <- \"datasets/contig_mapping_240621.csv\" # Path to mapping file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "177be20b-591f-4066-bc53-3466819d0bc5",
   "metadata": {},
   "source": [
    "Imports a vcf file and returns the annotated data frame\n",
    "vcf_files Path to vcf file\n",
    "ontig_mapping Table with mapping of contig to genome ID\n",
    "gff_df Dataframe containing the genome annotation e.g. produced by Prokka\n",
    "Returns A dataframe containing annotation fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "faa0a3b5-775c-4731-80a2-15f76081ad5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfToDataframe <- function(vcf_files, contig_mapping = contig_mapping, gff_df = gff_df) {\n",
    "  res <- list()\n",
    "  for (vcf.file in vcf_files) {\n",
    "    message(vcf.file)\n",
    "    vcf.content <- vcfR::read.vcfR(vcf.file, verbose = FALSE)\n",
    "    vcf.fix <- as.data.frame(vcf.content@fix) # contains chr, position and substitution informations\n",
    "    vcf.info <- vcfR::INFO2df(vcf.content) # get INFO field, contains DP, AF informations\n",
    "    if(nrow(vcf.fix) > 0) { # there are variants\n",
    "      dat <-  as.data.frame(cbind(vcf.fix[,c(1, 2, 4, 5, 6)], vcf.info[,c(1, 2)]))\n",
    "      dat$majorAF <- sapply(dat$AF, minorAfToMajorAf)\n",
    "      dat$genome <- contig_mapping[match(dat$CHROM, contig_mapping$contig),]$genome\n",
    "      dat$genome_hr <- translateGenomeIdToFullName(tolower(dat$genome))\n",
    "      dat$mouse.id <-  substr(tools::file_path_sans_ext(basename(vcf.file)), 1, 4)\n",
    "      dat$mouse.group <- translateMouseIdToTreatmentGroup(dat$mouse.id)\n",
    "      dat$day <- as.integer(substr(basename(vcf.file), 6, 7))\n",
    "      dat$phase <- binDaysByPhase(as.numeric(as.matrix(dat$day)))\n",
    "      dat$phase_num <- binDaysByPhaseGroup(dat$day)\n",
    "      dat$dp <- as.numeric(as.matrix(vcf.info$DP))\n",
    "      # annotate overlap with a ORF\n",
    "      dt.gff <- data.table(start = gff_df$start, end = gff_df$end,\n",
    "                           chr = as.character(as.matrix(gff_df$chr)), feature = gff_df$product)\n",
    "      colnames(dat)[1:2] <- c(\"chr\", \"start\")\n",
    "      dat$start <- as.integer(as.matrix(dat$start))\n",
    "      dat$chr <- as.character(as.matrix(dat$chr))\n",
    "      dat$end <- dat$start\n",
    "      dat2 <- as.data.table(dat)\n",
    "      setkey(dt.gff, chr, start, end)\n",
    "      annotated <- foverlaps(dat2, dt.gff, type = \"within\", mult = \"first\")\n",
    "      res[[tools::file_path_sans_ext(basename(vcf.file))]] <- annotated\n",
    "    } else{\n",
    "      message(paste(\"Skipping\", file))\n",
    "    }\n",
    "  }\n",
    "  df <- as.data.frame(do.call(rbind, res)) # merge list to df\n",
    "  return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "84547216-1e9b-441e-9934-6f82e7d0df75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>41101</li><li>7</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 41101\n",
       "\\item 7\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 41101\n",
       "2. 7\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 41101     7"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 7</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>start</th><th scope=col>end</th><th scope=col>type</th><th scope=col>gene</th><th scope=col>product</th><th scope=col>chr</th><th scope=col>genome</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>3</th><td> 403</td><td> 675</td><td>CDS</td><td>NA</td><td>type II toxin-antitoxin system YafQ family toxin</td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "\t<tr><th scope=row>5</th><td> 668</td><td> 910</td><td>CDS</td><td>NA</td><td>hypothetical protein                            </td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "\t<tr><th scope=row>7</th><td>1284</td><td>2432</td><td>CDS</td><td>NA</td><td>site-specific integrase                         </td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "\t<tr><th scope=row>9</th><td>2562</td><td>3911</td><td>CDS</td><td>NA</td><td>hypothetical protein                            </td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "\t<tr><th scope=row>11</th><td>3914</td><td>4321</td><td>CDS</td><td>NA</td><td>hypothetical protein                            </td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "\t<tr><th scope=row>13</th><td>4644</td><td>5006</td><td>CDS</td><td>NA</td><td>helix-turn-helix domain-containing protein      </td><td>Acutalibacter_muris_KB18</td><td>Acutalibacter_muris</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 7\n",
       "\\begin{tabular}{r|lllllll}\n",
       "  & start & end & type & gene & product & chr & genome\\\\\n",
       "  & <int> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t3 &  403 &  675 & CDS & NA & type II toxin-antitoxin system YafQ family toxin & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\t5 &  668 &  910 & CDS & NA & hypothetical protein                             & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\t7 & 1284 & 2432 & CDS & NA & site-specific integrase                          & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\t9 & 2562 & 3911 & CDS & NA & hypothetical protein                             & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\t11 & 3914 & 4321 & CDS & NA & hypothetical protein                             & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\t13 & 4644 & 5006 & CDS & NA & helix-turn-helix domain-containing protein       & Acutalibacter\\_muris\\_KB18 & Acutalibacter\\_muris\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 7\n",
       "\n",
       "| <!--/--> | start &lt;int&gt; | end &lt;int&gt; | type &lt;chr&gt; | gene &lt;chr&gt; | product &lt;chr&gt; | chr &lt;chr&gt; | genome &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| 3 |  403 |  675 | CDS | NA | type II toxin-antitoxin system YafQ family toxin | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "| 5 |  668 |  910 | CDS | NA | hypothetical protein                             | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "| 7 | 1284 | 2432 | CDS | NA | site-specific integrase                          | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "| 9 | 2562 | 3911 | CDS | NA | hypothetical protein                             | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "| 11 | 3914 | 4321 | CDS | NA | hypothetical protein                             | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "| 13 | 4644 | 5006 | CDS | NA | helix-turn-helix domain-containing protein       | Acutalibacter_muris_KB18 | Acutalibacter_muris |\n",
       "\n"
      ],
      "text/plain": [
       "   start end  type gene product                                         \n",
       "3   403   675 CDS  NA   type II toxin-antitoxin system YafQ family toxin\n",
       "5   668   910 CDS  NA   hypothetical protein                            \n",
       "7  1284  2432 CDS  NA   site-specific integrase                         \n",
       "9  2562  3911 CDS  NA   hypothetical protein                            \n",
       "11 3914  4321 CDS  NA   hypothetical protein                            \n",
       "13 4644  5006 CDS  NA   helix-turn-helix domain-containing protein      \n",
       "   chr                      genome             \n",
       "3  Acutalibacter_muris_KB18 Acutalibacter_muris\n",
       "5  Acutalibacter_muris_KB18 Acutalibacter_muris\n",
       "7  Acutalibacter_muris_KB18 Acutalibacter_muris\n",
       "9  Acutalibacter_muris_KB18 Acutalibacter_muris\n",
       "11 Acutalibacter_muris_KB18 Acutalibacter_muris\n",
       "13 Acutalibacter_muris_KB18 Acutalibacter_muris"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gff_df <- NULL\n",
    "for (gff.file in gff_files) {\n",
    "\n",
    "  gff <- rtracklayer::readGFF(gff.file)\n",
    "  relevant <- data.frame(start = gff$start, end = gff$end, type = as.character(as.matrix(gff$type)), gene = as.character(as.matrix(gff$gene)), product = as.character(as.matrix(gff$product)), chr = as.character(as.matrix(gff$seqid)))\n",
    "  relevant$genome <-  substr(basename(gff.file), 1, nchar(basename(gff.file)) - 5)\n",
    "  gff_df <- rbind(gff_df, relevant)\n",
    "  gff_df <- gff_df[which( gff_df$type== \"CDS\"),]\n",
    "}\n",
    "\n",
    "dim(gff_df)\n",
    "head(gff_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "460cc34c-fcae-49fd-a596-62e47aeddd0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 2</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>contig</th><th scope=col>genome</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Akkermansia_muciniphila_YL44</td><td>YL44</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Acutalibacter_muris_KB18    </td><td>KB18</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Bifidobacterium_animalis_YL2</td><td>YL2 </td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Bacteroides_caecimuris_I48  </td><td>I48 </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Blautia_coccoides_YL58      </td><td>YL58</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Clostridium_innocuum_I46    </td><td>I46 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 2\n",
       "\\begin{tabular}{r|ll}\n",
       "  & contig & genome\\\\\n",
       "  & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & Akkermansia\\_muciniphila\\_YL44 & YL44\\\\\n",
       "\t2 & Acutalibacter\\_muris\\_KB18     & KB18\\\\\n",
       "\t3 & Bifidobacterium\\_animalis\\_YL2 & YL2 \\\\\n",
       "\t4 & Bacteroides\\_caecimuris\\_I48   & I48 \\\\\n",
       "\t5 & Blautia\\_coccoides\\_YL58       & YL58\\\\\n",
       "\t6 & Clostridium\\_innocuum\\_I46     & I46 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 2\n",
       "\n",
       "| <!--/--> | contig &lt;chr&gt; | genome &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| 1 | Akkermansia_muciniphila_YL44 | YL44 |\n",
       "| 2 | Acutalibacter_muris_KB18     | KB18 |\n",
       "| 3 | Bifidobacterium_animalis_YL2 | YL2  |\n",
       "| 4 | Bacteroides_caecimuris_I48   | I48  |\n",
       "| 5 | Blautia_coccoides_YL58       | YL58 |\n",
       "| 6 | Clostridium_innocuum_I46     | I46  |\n",
       "\n"
      ],
      "text/plain": [
       "  contig                       genome\n",
       "1 Akkermansia_muciniphila_YL44 YL44  \n",
       "2 Acutalibacter_muris_KB18     KB18  \n",
       "3 Bifidobacterium_animalis_YL2 YL2   \n",
       "4 Bacteroides_caecimuris_I48   I48   \n",
       "5 Blautia_coccoides_YL58       YL58  \n",
       "6 Clostridium_innocuum_I46     I46   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "contig_mapping <- read.csv2(contig_mapping_path, sep = \";\",\n",
    "                            header = T, stringsAsFactors = F)\n",
    "head(contig_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aeb1a9e0-165b-440f-bfde-b5fe24ee3763",
   "metadata": {},
   "outputs": [],
   "source": [
    "minorAfToMajorAf <- function(x) if (x > 0) max(1 - x, x) else NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b57f0574-0a6a-41e5-b97d-100ea66ac3ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "pipeline_data/vcf/vcf_wgs_130721/1681d00_S56.vcf\n",
      "\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1681d14_S57.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1681d30_S58.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1681d49_S59.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1681d63_S60.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1681d79_S61.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d00_S47.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d04_S48.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d09_S49.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d14_S50.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d18_S51.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d23_S52.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d30_S53.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d37_S54.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d44_S55.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d49_S56.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d53_S57.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d58_S58.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d63_S59.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d67_S60.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d72_S61.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1683d79_S62.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d00_S62.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d14_S63.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d30_S64.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d49_S65.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d63_S66.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1684d79_S67.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d00_S69.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d14_S70.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d30_S71.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d49_S72.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d63_S73.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1686d79_S74.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d00_S1.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d04_S2.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d09_S3.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d14_S4.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d18_S5.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d23_S6.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d30_S7.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d37_S8.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d44_S9.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d49_S10.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d53_S11.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d58_S12.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d63_S13.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d67_S14.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d72_S15.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1688d79_S16.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d00_S75.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d14_S76.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d30_S77.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d49_S78.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d63_S79.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1690d79_S80.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d00_S17.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d04_S18.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d09_S19.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d14_S20.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d18_S21.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d23_S22.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d30_S23.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d37_S24.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d44_S25.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d49_S26.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d53_S27.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d58_S28.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d63_S29.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d67_S30.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d72_S31.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1692d79_S32.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d00_S81.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d14_S82.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d30_S83.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d49_S84.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d63_S85.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1693d79_S86.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d00_S87.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d14_S88.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d30_S89.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d49_S90.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d63_S91.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1694d79_S43.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d00_S50.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d14_S51.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d30_S52.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d49_S53.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d63_S54.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1697d79_S55.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d00_S44.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d14_S45.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d30_S46.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d49_S47.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d63_S48.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1698d79_S49.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d00_S33.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d04_S34.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d09_S35.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d14_S36.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d18_S37.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d23_S38.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d30_S39.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d37_S40.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d44_S41.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d49_S42.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d53_S63.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d58_S64.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d63_S65.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d67_S66.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d72_S67.vcf\n",
      "\n",
      "pipeline_data/vcf/vcf_wgs_130721/1699d79_S68.vcf\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 22</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>chr</th><th scope=col>feature</th><th scope=col>POS</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>QUAL</th><th scope=col>DP</th><th scope=col>AF</th><th scope=col>majorAF</th><th scope=col>genome</th><th scope=col>⋯</th><th scope=col>mouse.group</th><th scope=col>day</th><th scope=col>phase</th><th scope=col>phase_num</th><th scope=col>dp</th><th scope=col>ref_size</th><th scope=col>alt_size</th><th scope=col>alteration</th><th scope=col>alteration_type</th><th scope=col>snp_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>Akkermansia_muciniphila_YL44</td><td>autotransporter-associated beta strand repeat-containing protein</td><td>239850</td><td>C</td><td>T</td><td>1402</td><td>129</td><td>0.379845</td><td>0.620155</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>129</td><td>1</td><td>1</td><td>C -&gt; T</td><td>SNP</td><td>239850-C-T</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>Akkermansia_muciniphila_YL44</td><td>autotransporter-associated beta strand repeat-containing protein</td><td>240383</td><td>A</td><td>G</td><td>1086</td><td>116</td><td>0.344828</td><td>0.655172</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>116</td><td>1</td><td>1</td><td>A -&gt; G</td><td>SNP</td><td>240383-A-G</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>Akkermansia_muciniphila_YL44</td><td>autotransporter-associated beta strand repeat-containing protein</td><td>241793</td><td>A</td><td>G</td><td> 116</td><td>129</td><td>0.046512</td><td>0.953488</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>129</td><td>1</td><td>1</td><td>A -&gt; G</td><td>SNP</td><td>241793-A-G</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>Akkermansia_muciniphila_YL44</td><td><span style=white-space:pre-wrap>PEP-CTERM sorting domain-containing protein                     </span></td><td>296139</td><td>C</td><td>T</td><td>1067</td><td>160</td><td>0.250000</td><td>0.750000</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>160</td><td>1</td><td>1</td><td>C -&gt; T</td><td>SNP</td><td>296139-C-T</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>Akkermansia_muciniphila_YL44</td><td><span style=white-space:pre-wrap>protein kinase                                                  </span></td><td>355082</td><td>A</td><td>G</td><td>1281</td><td>130</td><td>0.353846</td><td>0.646154</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>130</td><td>1</td><td>1</td><td>A -&gt; G</td><td>SNP</td><td>355082-A-G</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>Akkermansia_muciniphila_YL44</td><td><span style=white-space:pre-wrap>protein kinase                                                  </span></td><td>355082</td><td>A</td><td>T</td><td> 950</td><td>130</td><td>0.276923</td><td>0.723077</td><td>YL44</td><td>⋯</td><td>Control</td><td>0</td><td>pre-treatment</td><td>1</td><td>130</td><td>1</td><td>1</td><td>A -&gt; T</td><td>SNP</td><td>355082-A-T</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 22\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & chr & feature & POS & REF & ALT & QUAL & DP & AF & majorAF & genome & ⋯ & mouse.group & day & phase & phase\\_num & dp & ref\\_size & alt\\_size & alteration & alteration\\_type & snp\\_id\\\\\n",
       "  & <chr> & <chr> & <int> & <chr> & <chr> & <int> & <int> & <dbl> & <dbl> & <chr> & ⋯ & <chr> & <int> & <chr> & <dbl> & <dbl> & <int> & <int> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & Akkermansia\\_muciniphila\\_YL44 & autotransporter-associated beta strand repeat-containing protein & 239850 & C & T & 1402 & 129 & 0.379845 & 0.620155 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 129 & 1 & 1 & C -> T & SNP & 239850-C-T\\\\\n",
       "\t2 & Akkermansia\\_muciniphila\\_YL44 & autotransporter-associated beta strand repeat-containing protein & 240383 & A & G & 1086 & 116 & 0.344828 & 0.655172 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 116 & 1 & 1 & A -> G & SNP & 240383-A-G\\\\\n",
       "\t3 & Akkermansia\\_muciniphila\\_YL44 & autotransporter-associated beta strand repeat-containing protein & 241793 & A & G &  116 & 129 & 0.046512 & 0.953488 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 129 & 1 & 1 & A -> G & SNP & 241793-A-G\\\\\n",
       "\t4 & Akkermansia\\_muciniphila\\_YL44 & PEP-CTERM sorting domain-containing protein                      & 296139 & C & T & 1067 & 160 & 0.250000 & 0.750000 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 160 & 1 & 1 & C -> T & SNP & 296139-C-T\\\\\n",
       "\t5 & Akkermansia\\_muciniphila\\_YL44 & protein kinase                                                   & 355082 & A & G & 1281 & 130 & 0.353846 & 0.646154 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 130 & 1 & 1 & A -> G & SNP & 355082-A-G\\\\\n",
       "\t6 & Akkermansia\\_muciniphila\\_YL44 & protein kinase                                                   & 355082 & A & T &  950 & 130 & 0.276923 & 0.723077 & YL44 & ⋯ & Control & 0 & pre-treatment & 1 & 130 & 1 & 1 & A -> T & SNP & 355082-A-T\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 22\n",
       "\n",
       "| <!--/--> | chr &lt;chr&gt; | feature &lt;chr&gt; | POS &lt;int&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | QUAL &lt;int&gt; | DP &lt;int&gt; | AF &lt;dbl&gt; | majorAF &lt;dbl&gt; | genome &lt;chr&gt; | ⋯ ⋯ | mouse.group &lt;chr&gt; | day &lt;int&gt; | phase &lt;chr&gt; | phase_num &lt;dbl&gt; | dp &lt;dbl&gt; | ref_size &lt;int&gt; | alt_size &lt;int&gt; | alteration &lt;chr&gt; | alteration_type &lt;chr&gt; | snp_id &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | Akkermansia_muciniphila_YL44 | autotransporter-associated beta strand repeat-containing protein | 239850 | C | T | 1402 | 129 | 0.379845 | 0.620155 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 129 | 1 | 1 | C -&gt; T | SNP | 239850-C-T |\n",
       "| 2 | Akkermansia_muciniphila_YL44 | autotransporter-associated beta strand repeat-containing protein | 240383 | A | G | 1086 | 116 | 0.344828 | 0.655172 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 116 | 1 | 1 | A -&gt; G | SNP | 240383-A-G |\n",
       "| 3 | Akkermansia_muciniphila_YL44 | autotransporter-associated beta strand repeat-containing protein | 241793 | A | G |  116 | 129 | 0.046512 | 0.953488 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 129 | 1 | 1 | A -&gt; G | SNP | 241793-A-G |\n",
       "| 4 | Akkermansia_muciniphila_YL44 | PEP-CTERM sorting domain-containing protein                      | 296139 | C | T | 1067 | 160 | 0.250000 | 0.750000 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 160 | 1 | 1 | C -&gt; T | SNP | 296139-C-T |\n",
       "| 5 | Akkermansia_muciniphila_YL44 | protein kinase                                                   | 355082 | A | G | 1281 | 130 | 0.353846 | 0.646154 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 130 | 1 | 1 | A -&gt; G | SNP | 355082-A-G |\n",
       "| 6 | Akkermansia_muciniphila_YL44 | protein kinase                                                   | 355082 | A | T |  950 | 130 | 0.276923 | 0.723077 | YL44 | ⋯ | Control | 0 | pre-treatment | 1 | 130 | 1 | 1 | A -&gt; T | SNP | 355082-A-T |\n",
       "\n"
      ],
      "text/plain": [
       "  chr                         \n",
       "1 Akkermansia_muciniphila_YL44\n",
       "2 Akkermansia_muciniphila_YL44\n",
       "3 Akkermansia_muciniphila_YL44\n",
       "4 Akkermansia_muciniphila_YL44\n",
       "5 Akkermansia_muciniphila_YL44\n",
       "6 Akkermansia_muciniphila_YL44\n",
       "  feature                                                          POS    REF\n",
       "1 autotransporter-associated beta strand repeat-containing protein 239850 C  \n",
       "2 autotransporter-associated beta strand repeat-containing protein 240383 A  \n",
       "3 autotransporter-associated beta strand repeat-containing protein 241793 A  \n",
       "4 PEP-CTERM sorting domain-containing protein                      296139 C  \n",
       "5 protein kinase                                                   355082 A  \n",
       "6 protein kinase                                                   355082 A  \n",
       "  ALT QUAL DP  AF       majorAF  genome ⋯ mouse.group day phase        \n",
       "1 T   1402 129 0.379845 0.620155 YL44   ⋯ Control     0   pre-treatment\n",
       "2 G   1086 116 0.344828 0.655172 YL44   ⋯ Control     0   pre-treatment\n",
       "3 G    116 129 0.046512 0.953488 YL44   ⋯ Control     0   pre-treatment\n",
       "4 T   1067 160 0.250000 0.750000 YL44   ⋯ Control     0   pre-treatment\n",
       "5 G   1281 130 0.353846 0.646154 YL44   ⋯ Control     0   pre-treatment\n",
       "6 T    950 130 0.276923 0.723077 YL44   ⋯ Control     0   pre-treatment\n",
       "  phase_num dp  ref_size alt_size alteration alteration_type snp_id    \n",
       "1 1         129 1        1        C -> T     SNP             239850-C-T\n",
       "2 1         116 1        1        A -> G     SNP             240383-A-G\n",
       "3 1         129 1        1        A -> G     SNP             241793-A-G\n",
       "4 1         160 1        1        C -> T     SNP             296139-C-T\n",
       "5 1         130 1        1        A -> G     SNP             355082-A-G\n",
       "6 1         130 1        1        A -> T     SNP             355082-A-T"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vcf_samples <- vcfToDataframe(vcf_files, contig_mapping, gff_df = gff_df)\n",
    "vcf_samples$feature <- as.character(as.matrix(vcf_samples$feature))\n",
    "vcf_samples[which(is.na(vcf_samples$feature)),]$feature <- \"outside ORFs\"\n",
    "vcf_samples$start <- NULL\n",
    "vcf_samples$end <- NULL\n",
    "vcf_samples$i.end <- NULL\n",
    "colnames(vcf_samples)[3] <- \"POS\"\n",
    "vcf_samples$ref_size <- nchar(as.character(as.matrix(vcf_samples$REF)))\n",
    "vcf_samples$alt_size <- nchar(as.character(as.matrix(vcf_samples$ALT)))\n",
    "vcf_samples$alteration <- paste(as.character(vcf_samples$REF), \"->\",as.character(vcf_samples$ALT))\n",
    "vcf_samples$alteration_type <- \"SNP\"\n",
    "vcf_samples$QUAL <- as.integer(as.matrix(vcf_samples$QUAL))\n",
    "vcf_samples <- vcf_samples[which(vcf_samples$QUAL >= qual_threshold),]\n",
    "vcf_samples$snp_id <- paste0(vcf_samples$POS, \"-\", vcf_samples$REF, \"-\", vcf_samples$ALT)\n",
    "head(vcf_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "978899cb-aba9-46e8-9b48-714b0ed020cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(vcf_samples, file = \"R/datasets/omm_wgs_rerun.rds\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
